src/Microsoft.Extensions.Configuration.AzureKeyVault/AzureKeyVaultConfigurationExtensions.cs

```diff
@@ -0,0 +1,152 @@
+            {
+                throw new ArgumentNullException(nameof(clientId));
+            }
+            if (clientSecret == null)
+            {
+                throw new ArgumentNullException(nameof(clientSecret));
+            }
+            KeyVaultClient.AuthenticationCallback callback =
+                (authority, resource, scope) => GetTokenFromClientSecret(authority, resource, scope, clientId, clientSecret);
+
+            return configurationBuilder.AddAzureKeyVault(vault, callback, filter);
+        }
+
+
+        /// <summary>
+        /// Adds an <see cref="IConfigurationProvider"/> that reads configuration values from the Azure KeyVault.
+        /// </summary>
+        /// <param name="configurationBuilder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="vault">Azure KeyVault uri.</param>
+        /// <param name="clientId">The application client id.</param>
+        /// <param name="certificate">The <see cref="X509Certificate2"/> to use for authentication.</param>
+        /// <param name="filter">The predicate to filter secret entries before loading value, <code>null</code> to load all.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder AddAzureKeyVault(
+            this IConfigurationBuilder configurationBuilder,
+            string vault,
+            string clientId,
+            X509Certificate2 certificate,
+            Func<SecretItem, bool> filter)
+        {
+            if (clientId == null)
+            {
+                throw new ArgumentNullException(nameof(clientId));
+            }
+            if (certificate == null)
+            {
+                throw new ArgumentNullException(nameof(certificate));
+            }
+            KeyVaultClient.AuthenticationCallback callback =
+                (authority, resource, scope) => GetTokenFromClientCertificate(authority, resource, scope, clientId, certificate);
+
+            return configurationBuilder.AddAzureKeyVault(vault, callback, filter);
+        }
+
+        /// <summary>
+        /// Adds an <see cref="IConfigurationProvider"/> that reads configuration values from the Azure KeyVault.
+        /// </summary>
+        /// <param name="configurationBuilder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="vault">Azure KeyVault uri.</param>
+        /// <param name="authenticationCallback">The <see cref="KeyVaultClient.AuthenticationCallback"/> to use for authentication.</param>
+        /// <param name="filter">The predicate to filter secret entries before loading value, <code>null</code> to load all.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder AddAzureKeyVault(
+            this IConfigurationBuilder configurationBuilder,
+            string vault,
+            KeyVaultClient.AuthenticationCallback authenticationCallback,
```

Comment:
Should the delegate be top level - e.g. `KeyVaaultAuthenticationCallback`? Unusual to see a nested type in an API.